Cabal/Distribution/PackageDescription/Parsec.hs

```diff
@@ -0,0 +1,566 @@
+        (die $ "Error Parsing: file \"" ++ fpath ++ "\" doesn't exist. Cannot continue.")
+    bs <- BS.readFile fpath
+    let (warnings, errors, result) = runParseResult (parser bs)
+    traverse_ (warn verbosity . showPWarning fpath) warnings
+    traverse_ (warn verbosity . showPError fpath) errors
+    case result of
+        Nothing -> die $ "Failing parsing \"" ++ fpath ++ "\"."
+        Just x  -> return x
+
+-- | Parse the given package file.
+readGenericPackageDescription :: Verbosity -> FilePath -> IO GenericPackageDescription
+readGenericPackageDescription = readAndParseFile parseGenericPackageDescription
+
+------------------------------------------------------------------------------
+-- | Parses the given file into a 'GenericPackageDescription'.
+--
+-- In Cabal 1.2 the syntax for package descriptions was changed to a format
+-- with sections and possibly indented property descriptions.
+--
+-- TODO: add lex warnings
+parseGenericPackageDescription :: BS.ByteString -> ParseResult GenericPackageDescription
+parseGenericPackageDescription bs = case readFields' bs of
+    Right (fs, lexWarnings) -> parseGenericPackageDescription' lexWarnings fs
+    -- TODO: better marshalling of errors
+    Left perr -> parseFatalFailure (Position 0 0) (show perr)
+
+runFieldParser :: FieldParser a -> [FieldLine Position] -> ParseResult a
+runFieldParser p ls = runFieldParser' pos p =<< fieldlinesToString pos ls
+  where
+    -- TODO: make per line lookup
+    pos = case ls of
+        []                     -> Position 0 0
+        (FieldLine pos' _ : _) -> pos'
+
+fieldlinesToBS :: [FieldLine ann] -> BS.ByteString
+fieldlinesToBS = BS.intercalate "\n" . map (\(FieldLine _ bs) -> bs)
+
+-- TODO: Take position  from FieldLine
+-- TODO: Take field name
+fieldlinesToString :: Position -> [FieldLine ann] -> ParseResult String
+fieldlinesToString pos fls =
+    let str = intercalate "\n" . map (\(FieldLine _ bs') -> fromUTF8BS bs') $ fls
+    in if '\xfffd' `elem` str
+        then str <$ parseWarning pos PWTUTF "Invalid UTF8 encoding"
+        else pure str
+
+runFieldParser' :: Position -> FieldParser a -> String -> ParseResult a
+runFieldParser' (Position row col) p str = case P.runParser p' [] "<field>" str of
+    Right (pok, ws) -> do
+        -- TODO: map pos
+        traverse_ (\(PWarning t pos w) -> parseWarning pos t w) ws
+        pure pok
+    Left err        -> do
+        let ppos = P.errorPos err
+        -- Positions start from 1:1, not 0:0
+        let epos = Position (row - 1 + P.sourceLine ppos) (col - 1 + P.sourceColumn ppos)
+        let msg = P.showErrorMessages
+                "or" "unknown parse error" "expecting" "unexpected" "end of input"
+                (P.errorMessages err)
+
+        parseFatalFailure epos $ msg ++ ": " ++ show str
+  where
+    p' = (,) <$ P.spaces <*> p <* P.spaces <* P.eof <*> P.getState
+
+data GPDS = Fields | Sections
+
+-- Note [Accumulating parser]
+parseGenericPackageDescription'
```

Comment:
Following the other naming convention, why isn't this called parsecGenericPackageDescription?